crates/ruff_python_semantic/src/cfg/graph.rs

```diff
@@ -0,0 +1,282 @@
+
+    /// Runs the core logic for the builder.
+    fn process_stmts(&mut self, stmts: &'stmt [Stmt]) {
+        let start = 0;
+        for stmt in stmts {
+            let cache_exit = self.exit();
+            match stmt {
+                Stmt::FunctionDef(_)
+                | Stmt::ClassDef(_)
+                | Stmt::Assign(_)
+                | Stmt::AugAssign(_)
+                | Stmt::AnnAssign(_)
+                | Stmt::TypeAlias(_)
+                | Stmt::Import(_)
+                | Stmt::ImportFrom(_)
+                | Stmt::Global(_)
+                | Stmt::Nonlocal(_)
+                | Stmt::Expr(_)
+                | Stmt::Pass(_)
+                | Stmt::Delete(_)
+                | Stmt::IpyEscapeCommand(_) => {}
+                // Loops
+                Stmt::While(_) => {}
+                Stmt::For(_) => {}
+
+                // Switch statements
+                Stmt::If(_) => {}
+                Stmt::Match(_) => {}
+
+                // Exception handling statements
+                Stmt::Try(_) => {}
+                Stmt::With(_) => {}
+
+                // Jumps
+                Stmt::Return(_) => {}
+                Stmt::Break(_) => {}
+                Stmt::Continue(_) => {}
+                Stmt::Raise(_) => {}
+
+                // An `assert` is a mixture of a switch and a jump.
+                Stmt::Assert(_) => {}
+            }
+            // Restore exit
+            self.update_exit(cache_exit);
+        }
+        // Push remaining statements
+        if start < stmts.len() {
+            self.set_stmts(&stmts[start..]);
+        }
+        // Add edge to exit if not already present
+        if self.cfg.blocks[self.current].out.is_empty() {
+            let edges = Edges::always(self.exit());
+            self.set_edges(edges);
+        }
+        self.move_to(self.exit());
+    }
+
+    /// Returns finished control flow graph
+    fn finish(self) -> CFG<'stmt> {
+        self.cfg
+    }
+
+    /// Current exit block, which may change during construction
+    fn exit(&self) -> BlockId {
+        self.exit
+    }
+
+    /// Point the current exit to block at provided index
+    fn update_exit(&mut self, new_exit: BlockId) {
+        self.exit = new_exit;
+    }
+
+    /// Moves current block to provided index
+    fn move_to(&mut self, block: BlockId) {
+        self.current = block;
+    }
+
+    /// Populates the current basic block with the given set of statements.
+    ///
+    /// This should only be called once on any given block.
+    fn set_stmts(&mut self, stmts: &'stmt [Stmt]) {
+        debug_assert!(
+            self.cfg.blocks[self.current].stmts.is_empty(),
+            "Attempting to set statements on an already populated basic block."
+        );
+        self.cfg.blocks[self.current].stmts = stmts;
+    }
+
+    /// Draws provided edges out of the current basic block.
+    ///
+    /// This should only be called once on any given block.
+    fn set_edges(&mut self, edges: Edges) {
```

Comment:
Nit: `set_current_block_stmts` and `set_current_block_edges` to make it clear that this only mutates the current block and not the state of the entire builder.